apache/metron
# METRON-1788 Batch profiler pull profile information from zookeeper

## Contributor Comments
An enhancement to the batch profiler to allow for pulling profile information from zookeeper.  Also added a 'timestampfield' option so the event time source field can be specified on the command line (using the processing time is not helpful for batch processing modes).  If existing command line and/or environment variables remain unchanged then the batch profiler will continue to read profile definitions from the file system.  

## Changes
* zookeeper and timestampfield command line options were added to the batch profiler (The batch profiler utilises the same Metron API to pull profile definitions from zookeeper as Stellar does).
* The wrapper script for batch profiler was updated to support the new options. To support this the wrapper script now passes command line options to the underlying program.
* The Readme file were updated to document the new options.

## Testing
Unit tests added for additional parameter handling.  A Metron dev cluster was manually spun up to test reading of profile information.  As the batch profiler now utilises the same API as Stellar does to read profile information,  it also leverages the same testing that is setup around the Stellar engine.  Documentation markdown rendering was checked utilising the Intellij Markdown module, as I was unable to get site-book building running on MacOS.
Travis CI testing run completed successfully on private repo.

### Manual Testing

1. Deploy a single VM Developer deployment.
1. Manually install Spark2 via Ambari.
1. Vagrant SSH to login to the developer box.
1. Source /etc/default/metron
1. Create the profile definition in a file located at $METRON_HOME/config/zookeeper/profiler.json. 
```
{
  "profiles": [
    {
      "profile": "hello-world",
      "onlyif":  "exists(ip_src_addr)",
      "foreach": "ip_src_addr",
      "init":    { "count": "0" },
      "update":  { "count": "count + 1" },
      "result":  "count"
    }
  ],
"timestampField": "timestamp"
}
```
1. `cd $METRON_HOME`
1.  Push the profile to zookeeper via `bin/zk_load_configs.sh -m PUSH -i config/zookeeper/ -z $ZOOKEEPER`
1.  Enable zookeeper profile reading via `export SPARK_PROFILER_USE_ZOOKEEPER=1`
1.  `bin/start_batch_profiler.sh`
1.  Scan the produced log entries for an entry similar to the following to confirm correct reading of a profile from Zookeeper
```
19/05/01 06:55:48 INFO BatchProfilerCLI: Utilising profile: org.apache.metron.common.configuration.profiler.ProfilerConfig@40e10ff8[profiles=[org.apache.metron.common.configuration.profiler.ProfileConfig@1433046b[profile=hello-world,foreach=ip_src_addr,onlyif=exists(ip_src_addr),init={count=0},update={count=count + 1},groupBy=[],result=ProfileResult{profileExpressions=ProfileResultExpressions{expression='count'}, triageExpressions=ProfileTriageExpressions{expressions={}}},expires=<null>]],timestampField=foo]
```

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:


### For code changes:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.
